File: src/leetcode/python/hard/MinimumWindowSubstring_76.py
class MinimumWindowSubstring_76(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        if len(s) < len(t):
            return ""
        import sys
        length = sys.maxsize
        required = [0] * 256
        found = [0] * 256
        cnt = 0
        for i in range(len(t)):
            required[ord(t[i])] += 1
        k = 0
        start = -1
        for j in range(len(s)):
            if required[ord(s[j])] == 0:
                continue
            found[ord(s[j])] += 1
            if found[ord(s[j])] <= required[ord(s[j])]:
                cnt += 1
            if cnt == len(t):
                while required[ord(s[k])] == 0 or found[ord(s[k])] > required[ord(s[k])]:
                    if found[ord(s[k])] > required[ord(s[k])]:
                        found[ord(s[k])] -= 1
                    k += 1
                if j - k + 1 < length:
                    length = j - k + 1
                    start = k
        return "" if length == sys.maxsize else s[start:start + length]

File: src/leetcode/java/MinimumWindowSubstring_76.java
package leetcode.java.hard;

public class MinimumWindowSubstring_76 {
  public String minWindow(String s, String t) {
    if (s.length() < t.length()) return "";
    char[] sc = s.toCharArray();
    char[] tc = t.toCharArray();
    int start = -1;
    int len = Integer.MAX_VALUE;
    int cnt = 0;
    int[] required = new int[256];
    int[] found = new int[256];
    for (char tch : tc) required[tch]++;
    for (int i = 0, j = 0; i < sc.length; i++) {
      if (required[sc[i]] == 0) continue;
      if (++found[sc[i]] <= required[sc[i]]) cnt++;
      if (cnt == tc.length) {
        while (required[sc[j]] == 0 || found[sc[j]] > required[sc[j]]) {
          if (found[sc[j]] > required[sc[j]]) found[sc[j]]--;
          j++;
        }
        if (i - j + 1 < len) {
          start = j;
          len = i - j + 1;
        }
      }
    }
    return len == Integer.MAX_VALUE ? "" : s.substring(start, start + len);
  }
}
